From 5e7b6f6336d98754d2059f71afb8883fa96323a3 Mon Sep 17 00:00:00 2001
From: Aniket Randive <arandive@qti.qualcomm.com>
Date: Mon, 8 Sep 2025 17:04:37 +0530
Subject: [PATCH 07/13] dmaengine: qcom: gpi: Add premature cancel support for
 CCU QUP

Add support for handling premature cancel scenarios in the GPI driver
for CCU QUP. This includes several improvements and fixes to ensure
proper cleanup and robustness during cancellation.

Changes include:

- Identify the last frame and set the NOP_LAST flag in the NOP TRE.
- Refactor gpi_nop_tre() to align with upstream coding style.
- Update the GPII cancellation sequence as per the CCU QUP HPG.
- Release mutex on SYNC_TRIGGER_EN command failure to avoid deadlocks.
- Release async buffers and reset GPII struct variables after a
  successful GPII reset.
- Fix incorrect argument passed to geni_i2c_gpi_unmap() and refine
  cleanup in the SYNC transfer path for CCU QUP.

Change-Id: I95dc7f16f1f948e7c21140d8396ca6d9c12b2fb5
Signed-off-by: Maramaina Naresh <mnaresh@qti.qualcomm.com>
Signed-off-by: Aniket Randive <arandive@qti.qualcomm.com>
---
 drivers/dma/qcom/gpi.c             | 73 ++++++++++++++------------
 drivers/i2c/busses/i2c-qcom-geni.c | 84 ++++++++++++++++++++----------
 drivers/soc/qcom/qcom_ccu_qup.c    |  4 ++
 3 files changed, 101 insertions(+), 60 deletions(-)

diff --git a/drivers/dma/qcom/gpi.c b/drivers/dma/qcom/gpi.c
index bffd65d5707a..b3ad568c7409 100644
--- a/drivers/dma/qcom/gpi.c
+++ b/drivers/dma/qcom/gpi.c
@@ -102,24 +102,6 @@
 /* Lock Unlock TRE */
 #define TRE_I2C_UNLOCK		BIT(8)
 
-/* NOP TRE */
-#define NOP_TRE_LINK_RX	BIT(11)
-#define NOP_TRE_BEI		BIT(10)
-#define NOP_TRE_IEOT		BIT(9)
-#define NOP_TRE_IEOB		BIT(8)
-#define NOP_TRE_NOP_LAST BIT(3)
-#define NOP_TRE_CH_LAST BIT(2)
-#define NOP_TRE_CCU_LAST BIT(1)
-#define NOP_TRE_CH		BIT(0)
-#define NOP_TRE_MASK(link_rx, bei, ieot, ieob, ch) \
-		(FIELD_PREP(NOP_TRE_LINK_RX, link_rx)	| \
-		FIELD_PREP(NOP_TRE_BEI, bei)	|\
-		FIELD_PREP(NOP_TRE_IEOT, ieot)	|\
-		FIELD_PREP(NOP_TRE_IEOB, ieob)	|\
-		FIELD_PREP(NOP_TRE_CH, ch))
-
-#define NOP_TRE	BIT(16)
-
 /* Register offsets from gpi-top */
 #define GPII_n_CH_k_CNTXT_0_OFFS(n, k, ccu)\
 	((ccu) ? (0x20000 + (0x1000 * (n)) + (0x80 * (k)))\
@@ -1447,8 +1429,13 @@ static void gpi_process_xfer_compl_event(struct gchan *gchan,
 	unsigned long flags;
 	u32 chid;
 
-	/* only process events on active channel */
-	if (unlikely(gchan->pm_state != ACTIVE_STATE)) {
+	/*
+	 * Process events only if the channel is active or if premature_cancel is set.
+	 * The premature_cancel flag is used during gpi_terminate_all to allow cleanup
+	 * even when the channel is no longer active.
+	 */
+	if (!gpii->premature_cancel && unlikely(gchan->pm_state != ACTIVE_STATE)) {
+ 
 		GPI_ERR_LOG(LOG_LVL_ERROR, gpii->gpi_dev->dev,
 			    "skipping processing event because ch @ %s state\n",
 			    TO_GPI_PM_STR(gchan->pm_state));
@@ -1952,28 +1939,29 @@ static void gpi_queue_xfer(struct gpii *gpii, struct gchan *gchan,
 static void gpi_nop_tre(struct gpii *gpii, struct gchan *gchan)
 {
 	struct gpi_ring *ring = &gchan->ch_ring;
-	struct gpi_tre *tre;
+	struct gpi_tre *tre = NULL;
+	struct gpi_tre *last_frame = NULL;
 	phys_addr_t local_rp, local_wp;
 	void *cntxt_rp;
-	u32 nop_mask, nop_tre;
 
 	local_rp = to_physical(ring, ring->rp);
 	local_wp = to_physical(ring, ring->wp);
 	cntxt_rp = ring->rp;
 
-	dev_dbg(gpii->gpi_dev->dev, "local_rp:0x%11llx, local_wp:0x%11llx\n",
-		local_rp, local_wp);
-
-	nop_mask = NOP_TRE_MASK(1, 0, 0, 0, 1);
-	nop_tre = NOP_TRE;
+	dev_dbg(gpii->gpi_dev->dev, "chid:%d, rp:0x%llx, wp:0x%llx\n",
+		gchan->chid, local_rp, local_wp);
 
 	while (local_rp != local_wp) {
 		tre = (struct gpi_tre *)cntxt_rp;
-		dev_dbg(gpii->gpi_dev->dev, "TRE: %08x %08x %08x %08x\n",
+		dev_dbg(gpii->gpi_dev->dev, "%s: TRE: %08x:%08x:%08x:%08x\n", __func__,
 			tre->dword[0], tre->dword[1], tre->dword[2],
 			tre->dword[3]);
-		tre->dword[3] &= nop_mask;
-		tre->dword[3] |= nop_tre;
+
+		if (tre->dword[3] & TRE_FLAGS_CCU_LAST)
+			last_frame = tre;
+
+		tre->dword[3] &= (TRE_FLAGS_CHAIN | TRE_FLAGS_LINK);
+		tre->dword[3] |= u32_encode_bits(TRE_TYPE_NOP, TRE_FLAGS_TYPE);
 		local_rp += ring->el_size;
 		cntxt_rp += ring->el_size;
 		if (cntxt_rp >= (ring->base + ring->len)) {
@@ -1981,6 +1969,17 @@ static void gpi_nop_tre(struct gpii *gpii, struct gchan *gchan)
 			local_rp = to_physical(ring, ring->base);
 		}
 	}
+
+	if (tre) {
+		tre->dword[3] |= u32_encode_bits(0, TRE_FLAGS_CHAIN);
+		tre->dword[3] |= u32_encode_bits(1, TRE_FLAGS_IEOB);
+		if (last_frame)
+			last_frame->dword[3] |= u32_encode_bits(1, TRE_FLAGS_NOP_LAST);
+
+		dev_dbg(gpii->gpi_dev->dev, "chid:%d, rp:0x%llx, wp:0x%llx, TRE: %08x:%08x:%08x:%08x\n",
+			gchan->chid, local_rp, local_wp, tre->dword[0], tre->dword[1],
+			tre->dword[2], tre->dword[3]);
+	}
 }
 
 /* reset and restart transfer channel */
@@ -1990,6 +1989,7 @@ static int gpi_terminate_all(struct dma_chan *chan)
 	struct gpii *gpii = gchan->gpii;
 	int schid, echid, i;
 	int ret = 0;
+	bool is_ccu_qup = gpii->gpi_dev->cfg.is_ccu_qup;
 
 	mutex_lock(&gpii->ctrl_lock);
 
@@ -2023,7 +2023,7 @@ static int gpi_terminate_all(struct dma_chan *chan)
 		ret = gpi_stop_chan(gchan);
 	}
 
-	if (gpii->premature_cancel) {
+	if (gpii->premature_cancel && ret == 0) {
 		for (i = schid; i < echid; i++) {
 			gchan = &gpii->gchan[i];
 			gpi_nop_tre(gpii, gchan);
@@ -2054,7 +2054,14 @@ static int gpi_terminate_all(struct dma_chan *chan)
 start_chan:
 	/* restart the channels */
 	for (i = schid; i < echid; i++) {
-		gchan = &gpii->gchan[i];
+		if (is_ccu_qup) {
+			/* Issue channel START command (1st for RX, followed by TX) */
+			u8 idx = echid - 1 - i;
+
+			gchan = &gpii->gchan[idx];
+		} else {
+			gchan = &gpii->gchan[i];
+		}
 
 		ret = gpi_start_chan(gchan);
 		if (ret) {
@@ -2070,7 +2077,7 @@ static int gpi_terminate_all(struct dma_chan *chan)
 		if (ret) {
 			dev_err(gpii->gpi_dev->dev, "Error with cmd:%s ret:%d\n",
 				TO_GPI_CMD_STR(GPI_CH_CMD_SYNC_TRIGGER_EN), ret);
-			return ret;
+			goto terminate_exit;
 		}
 	}
 
diff --git a/drivers/i2c/busses/i2c-qcom-geni.c b/drivers/i2c/busses/i2c-qcom-geni.c
index 4461f4988be1..e6ff6b883bd5 100644
--- a/drivers/i2c/busses/i2c-qcom-geni.c
+++ b/drivers/i2c/busses/i2c-qcom-geni.c
@@ -307,6 +307,8 @@ static const struct geni_i2c_clk_fld geni_ccu_i2c_clk_map[] = {
 static int setup_gpi_dma(struct geni_i2c_dev *gi2c, int idx);
 static void release_gpi_dma(struct geni_i2c_dev *gi2c);
 static void i2c_gpi_cb_result(void *cb, const struct dmaengine_result *result);
+static void geni_i2c_gpi_release_ccu_frame(struct geni_i2c_dev *gi2c,
+					   struct ccu_frame_info *frame);
 
 static int log_level = -1;
 static ssize_t log_level_show(struct device *dev,
@@ -627,7 +629,7 @@ static int geni_i2c_configure_ccu_mode(struct i2c_adapter *adap, int gpii_no,
 	}
 	idx = (gpii_no % CCU_MAX_GPII);
 
-	if ((ch_type != 1 && ch_type != 2) || idx < 0) {
+	if (ch_type < 0 || ch_type > 2 || idx < 0) {
 		dev_err(gi2c->se.dev, "%s: Invalid request, ch_type:%d, mode:%d gpii_no:%d\n",
 			__func__, ch_type, mode, gpii_no);
 		return -EINVAL;
@@ -649,7 +651,6 @@ static int geni_i2c_configure_ccu_mode(struct i2c_adapter *adap, int gpii_no,
 		ccu_gpi_configure_mode(gi2c->tx_c[idx], mode, ch_type);
 		ccu_gpi_configure_mode(gi2c->rx_c[idx], mode, ch_type);
 		dmaengine_terminate_sync(gi2c->tx_c[idx]);
-		dmaengine_terminate_sync(gi2c->rx_c[idx]);
 	}
 
 	gi2c->ccu.gpi[idx].gpii_mode_change = false;
@@ -667,7 +668,7 @@ static int geni_i2c_configure_ccu_mode(struct i2c_adapter *adap, int gpii_no,
  * Return 0 on success and error on failure.
  */
 static int geni_i2c_gpii_reset(struct i2c_adapter *adap, int gpii_no,
-			bool premature_cancel)
+			       bool premature_cancel)
 {
 	struct geni_i2c_dev *gi2c;
 	int idx = gpii_no % CCU_MAX_GPII;
@@ -684,24 +685,28 @@ static int geni_i2c_gpii_reset(struct i2c_adapter *adap, int gpii_no,
 		}
 	}
 
-	if (premature_cancel)
+	if (premature_cancel) {
+		if (idx == REGULAR_GPII) {
+			dev_err(gi2c->se.dev, "%s: premature cancel is invalid for a regular channel:%d\n",
+				__func__, gpii_no);
+			return -EINVAL;
+		}
 		ccu_gpi_set_premature_cancel_flag(gi2c->tx_c[idx]);
+	}
 
 	if (gi2c->tx_c[idx]) {
-		ret = dmaengine_terminate_all(gi2c->tx_c[idx]);
+		ret = dmaengine_terminate_sync(gi2c->tx_c[idx]);
 		if (ret) {
-			dev_err(gi2c->se.dev, "%s: gpi terminate failed for tx_c[%d]ret:%d\n",
-				__func__, ret, idx);
+			dev_err(gi2c->se.dev, "%s: gpi terminate failed for tx_c[%d] ret:%d\n",
+				__func__, idx, ret);
 			return ret;
 		}
-	}
 
-	if (gi2c->rx_c[idx]) {
-		ret = dmaengine_terminate_all(gi2c->rx_c[idx]);
-		if (ret) {
-			dev_err(gi2c->se.dev, "%s: gpi terminate failed for rx_c[%d]ret:%d\n",
-				__func__, ret, idx);
-			return ret;
+		for (u8 i = 0; i < MAX_NUM_FRAMES; i++) {
+			if (gi2c->ccu.gpi[idx].frame[i].async)
+				geni_i2c_gpi_release_ccu_frame(gi2c, &gi2c->ccu.gpi[idx].frame[i]);
+			gi2c->ccu.gpi[idx].released_cnt = 0;
+			gi2c->ccu.gpi[idx].recvd_cnt = 0;
 		}
 	}
 
@@ -1114,7 +1119,6 @@ static int geni_i2c_gpi_nop_tre(struct geni_i2c_dev *gi2c, struct dma_slave_conf
 
 err:
 	dev_err(gi2c->se.dev, "%s: GPI transfer failed: %d\n", __func__, ret);
-	dmaengine_terminate_sync(gi2c->rx_c[idx]);
 	dmaengine_terminate_sync(gi2c->tx_c[idx]);
 
 err_submit:
@@ -1167,19 +1171,16 @@ static int geni_i2c_tx_one_msg(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 }
 
 /*
- * geni_i2c_release_buf - This function is used to free the dma buffer
+ * geni_i2c_gpi_release_ccu_frame - Release a CCU frame
  *
  * @gi2c: Handle of the I2C device.
- * @idx: indicate the gpii no
+ * @frame: Pointer to the CCU frame to be released.
  */
-static void  geni_i2c_release_buf(struct geni_i2c_dev *gi2c, int idx)
+static void geni_i2c_gpi_release_ccu_frame(struct geni_i2c_dev *gi2c, struct ccu_frame_info *frame)
 {
-	int i, msgs, frame_idx;
-	struct ccu_frame_info *frame;
 	enum dma_data_direction map_dirn;
+	int i, msgs;
 
-	frame_idx = gi2c->ccu.gpi[idx].released_cnt;
-	frame = &gi2c->ccu.gpi[idx].frame[frame_idx];
 	msgs = frame->msgs;
 
 	for (i = 0; i < msgs; i++) {
@@ -1203,6 +1204,21 @@ static void  geni_i2c_release_buf(struct geni_i2c_dev *gi2c, int idx)
 		frame->msg[i].addr = 0;
 	}
 	memset(frame, 0, sizeof(*frame));
+}
+
+/*
+ * geni_i2c_release_buf - This function is used to free the dma buffer
+ *
+ * @gi2c: Handle of the I2C device.
+ * @idx: indicate the gpii no
+ */
+static void  geni_i2c_release_buf(struct geni_i2c_dev *gi2c, int idx)
+{
+	int frame_idx;
+
+	frame_idx = gi2c->ccu.gpi[idx].released_cnt;
+
+	geni_i2c_gpi_release_ccu_frame(gi2c, &gi2c->ccu.gpi[idx].frame[frame_idx]);
 
 	if (frame_idx >= MAX_NUM_FRAMES -1)
 		gi2c->ccu.gpi[idx].released_cnt = 0;
@@ -1263,12 +1279,19 @@ static void geni_i2c_gpi_unmap(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 {
 	if (tx_buf) {
 		dma_unmap_single(gi2c->se.dev->parent, tx_addr, msg->len, DMA_TO_DEVICE);
-		i2c_put_dma_safe_msg_buf(tx_buf, msg, !gi2c->err);
+		if (gi2c->is_ccu_qup) {
+			kfree(tx_buf);
+			tx_buf = NULL;
+		} else {
+			i2c_put_dma_safe_msg_buf(tx_buf, msg, !gi2c->err);
+		}
 	}
 
 	if (rx_buf) {
 		dma_unmap_single(gi2c->se.dev->parent, rx_addr, msg->len, DMA_FROM_DEVICE);
-		if (!gi2c->is_ccu_qup)
+		if (gi2c->is_ccu_qup)
+			rx_buf = NULL;
+		else
 			i2c_put_dma_safe_msg_buf(rx_buf, msg, !gi2c->err);
 	}
 }
@@ -1649,12 +1672,16 @@ static int geni_i2c_gpi_xfer(struct geni_i2c_dev *gi2c, struct i2c_msg msgs[], i
 						msgs[i].len, DMA_FROM_DEVICE);
 		}
 
-		geni_i2c_gpi_unmap(gi2c, &msgs[i], tx_buf, tx_addr, rx_buf, rx_addr);
 		if (gi2c->is_ccu_qup) {
+			geni_i2c_gpi_unmap(gi2c, &msgs[i], frame->msg[i].tx_buf,
+					   frame->msg[i].addr, rx_buf, rx_addr);
+
 			if (gi2c->ccu.gpi[idx].released_cnt >= MAX_NUM_FRAMES - 1)
 				gi2c->ccu.gpi[idx].released_cnt = 0;
 			else
 				gi2c->ccu.gpi[idx].released_cnt++;
+		} else {
+			geni_i2c_gpi_unmap(gi2c, &msgs[i], tx_buf, tx_addr, rx_buf, rx_addr);
 		}
 next_msg:
 		if (gi2c->is_ccu_qup) {
@@ -1686,9 +1713,12 @@ static int geni_i2c_gpi_xfer(struct geni_i2c_dev *gi2c, struct i2c_msg msgs[], i
 		kfree(gi2c->ccu.msg[i]);
 		gi2c->ccu.msg[i] = NULL;
 	}
-	dmaengine_terminate_sync(gi2c->rx_c[idx]);
 	dmaengine_terminate_sync(gi2c->tx_c[idx]);
-	geni_i2c_gpi_unmap(gi2c, &msgs[i], tx_buf, tx_addr, rx_buf, rx_addr);
+	if (gi2c->is_ccu_qup)
+		geni_i2c_gpi_unmap(gi2c, &msgs[i], frame->msg[i].tx_buf,
+				   frame->msg[i].addr, rx_buf, rx_addr);
+	else
+		geni_i2c_gpi_unmap(gi2c, &msgs[i], tx_buf, tx_addr, rx_buf, rx_addr);
 	return ret;
 }
 
diff --git a/drivers/soc/qcom/qcom_ccu_qup.c b/drivers/soc/qcom/qcom_ccu_qup.c
index 35dd4be5efc8..17c31b773f22 100644
--- a/drivers/soc/qcom/qcom_ccu_qup.c
+++ b/drivers/soc/qcom/qcom_ccu_qup.c
@@ -448,6 +448,10 @@ int ccu_gpii_reset(struct i2c_adapter *adap, int gpii_no, bool partial,
 			return ret;
 	} else {
 		for (i = 0; i < CCU_MAX_GPII_PER_SE; i++) {
+			/* Skip premature cancel for regular channel */
+			if (premature_cancel && i == 0)
+				continue;
+
 			ret = ccu_reset(adap, i, premature_cancel);
 			if (ret)
 				return ret;
-- 
2.34.1

